### PR TITLE
Preserve integer dtype when warping an image using `NearestNeighbour` interpolation

### LINKED ISSUE #4572: `sct_register_multimodal` is changing the data type when using nearest neighbours interpolation
## Description

I needed to use `sct_register_multimodal -identity 1 -x nn` to put segmentation masks in the same orientation/space as my images and I noticed that the function was changing the type of my output label. Is it normal ?

## Example

The type of my input label is `INT16` before using the function and the type of my image is `FLOAT64`.

After running
```
sct_register_multimodal -i labelsTr/sub-amuAL_T1w_label.nii.gz -d imagesTr/sub-amuAL_T1w.nii.gz -o reg/sub-amuAL_T1w.nii.gz -identity 1 -x nn
```

The type of my output label is `FLOAT64` even though I used nearest neighbour interpolation. Is it what I should expect ?

### PR DESCRIPTION
**Pull request by joshuacwnewton**
## Description

This will preserve integer images when `-x nn` is used (which I've explicitly referenced in my test).

## Linked issues

Fixes #4572.

### CODE REVIEW

**Comment by joshuacwnewton** [comment]
It looks like the tests here are only failing because of #4679.

File: spinalcordtoolbox/scripts/sct_apply_transfo.py [hunk 1]
```diff
@@ -109,6 +111,44 @@ def get_parser():
...
+             ['-r', fname_reference] +
+             interpolation_args,  # list of ['-n', interp_type]
+             verbose=verbose,
+             is_sct_binary=True)
+    # Preserve input datatype if output data can be safely cast to it
+    # Note: This is mostly relevant for integer input images + NearestNeighbour
+    #       interpolation, since ANTs will automatically cast to float, but
+    #       we usually want to preserve the integer datatype.
+    im_out = Image(fname_output)
+    if np.array_equal(im_out.data, im_out.data.astype(dtype_in)):
```
Comment (mguaypaq) [inline spinalcordtoolbox/scripts/sct_apply_transfo.py:145]: I would have expected this `if` statement to do a static check for certain arguments in the call to `isct_antsAppleTransforms` (something like `'nn' in interpolation_args`), rather than a dynamic check of whether the actual data values are affected by the casting operation. Is there a case that's better handled this way?

I ask because my experience is that having the output _type_ change based on the output _values_ can lead to very hard to debug corner cases.

**Review by mguaypaq — COMMENTED** [review]
Thanks for adding the test, and congrats on adding the long-desired wrapper for `isct_antsApplyTransforms`! I do have one question about the approach, but otherwise this looks good.

File: spinalcordtoolbox/scripts/sct_apply_transfo.py (hunk 1 above)
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_apply_transfo.py:145]: Thank you for the feedback! It definitely makes sense to be explicit about the cases we wish to apply this logic too.

I think the reason I wrote it this way was because I figured it was a safe operation to apply blindly (since we still ensure array equality), and we're applying this operation to a file that we generate (meaning the file is kind of still within "our domain")? And I figured it might help us catch other instances of type changing that we can't think of on the spot (e.g. the user supplies a float32 file or something because they're keen on saving disk space, but ANTs outputs a float64 file by default, etc.)

But, I'm alright with reverting my logic and explicitly defining the cases we want to apply this logic to (as to match the scope of the original issue).

File: spinalcordtoolbox/scripts/sct_apply_transfo.py [hunk 2]
```diff
@@ -109,6 +111,43 @@ def get_parser():
...
+             ['-r', fname_reference] +
+             interpolation_args,  # list of ['-n', interp_type]
+             verbose=verbose,
+             is_sct_binary=True)
+    # Preserve integer datatype when interpolation is NearestNeighour to
+    # counter ANTs behavior of always outputting float64 images
+    im_out = Image(fname_output)
+    dtype_in = Image(fname_input).data.dtype
+    input_is_int = issubclass(dtype_in.type, np.integer)
+    if input_is_int and 'NearestNeighbor' in interpolation_args:
```
Comment (mguaypaq) [inline spinalcordtoolbox/scripts/sct_apply_transfo.py:144]: It might be even simpler to just have the rule that NearestNeighbor interpolation outputs the same dtype as the input, regardless of integer/float/etc.? But I'm happy either way.
```suggestion
    if 'NearestNeighbor' in interpolation_args:
```

**Review by mguaypaq — APPROVED** [review]
Thanks for the update! And I do think your reasoning was very reasonable. But I still prefer this more static approach to dtypes.
